core/src/main/kotlin/org/evomaster/core/search/gene/MapGene.kt

```diff
@@ -100,4 +112,72 @@ class MapGene<T>(
         else listOf(this).plus(elements.flatMap { g -> g.flatView(excludePredicate) })
     }
 
+    override fun archiveMutation(randomness: Randomness, allGenes: List<Gene>, apc: AdaptiveParameterControl, selection: GeneMutationSelectionMethod, impact: GeneImpact?, geneReference: String, archiveMutator: ArchiveMutator, evi: EvaluatedIndividual<*>, targets: Set<Int>) {
+        var add = elements.isEmpty()
+        var delete = (elements.size == maxSize)
+
+        val fmodifySize = if (add || delete) false
+        else if (archiveMutator.applyArchiveSelection()
+                && impact != null
+                && impact is MapGeneImpact
+                && impact.sizeImpact.noImprovement.any { it.value < 2 } //if there is recent improvement by manipulating size
+        ){
+            randomness.nextBoolean(0.3)
+        }else {
+            randomness.nextBoolean(MODIFY_SIZE)
+        }
+        if (fmodifySize){
+            val p = randomness.nextBoolean()
+            add = add || p
+            delete = delete || !p
+        }
+
+        if (add && add == delete)
+            log.warn("add and delete an element cannot happen in a mutation, and size of elements: {} and maxSize: {}", elements.size, maxSize)
```

Comment:
please put parentheses. When I read code like that, I am never sure if it is `(add && add) == delete`, or `add && (add == delete)`